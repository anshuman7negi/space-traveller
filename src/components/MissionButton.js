import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { joinMission, leaveMission } from '../redux/missions/missionSlice';

const MissionButton = ({ id, reserved }) => {
  const dispatch = useDispatch();

  const click = (id) => {
    if (reserved) dispatch(joinMission(id));
    dispatch(leaveMission(id));
  };

  return (
    <button
      type="button"
      className={`${
        reserved
          ? 'text-blue-500 bg-transparent cursor-pointer mt-3 rounded-full px-[1.2rem] py-2 font-medium outline outline-red-500'
          : 'text-gray-500 bg-transparent cursor-pointer mt-3 rounded px-[1.2rem] py-2 font-medium outline outline-gray-400'
      }`}
      onClick={() => click(id)}
    >
      {reserved ? 'Leave missions' : 'Join missions'}
    </button>
  );
};

MissionButton.propTypes = {
  id: PropTypes.string.isRequired,
  reserved: PropTypes.bool.isRequired,
};

export default MissionButton;
