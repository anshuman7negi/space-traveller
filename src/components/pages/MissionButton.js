const MissionButton = () => {
    <button type="submit">Button</button>
};

export default MissionButton;
