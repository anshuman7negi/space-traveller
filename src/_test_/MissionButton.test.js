import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useDispatch } from 'react-redux';
import MissionButton from '../../src/components/MissionButton';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('MissionButton Component', () => {
  const mockDispatch = jest.fn();

  beforeEach(() => {
    useDispatch.mockReturnValue(mockDispatch);
  });

  test('renders "Join missions" button by default', () => {
    const { getByText } = render(<MissionButton id="123" reserved={false} />);
    const joinButton = getByText('Join missions');
    expect(joinButton).toBeInTheDocument();
  });

  test('renders "Leave missions" button when reserved is true', () => {
    const { getByText } = render(<MissionButton id="123" reserved={true} />);
    const leaveButton = getByText('Leave missions');
    expect(leaveButton).toBeInTheDocument();
  });

  test('dispatches joinMission action when clicking on "Join missions" button', () => {
    const { getByText } = render(<MissionButton id="123" reserved={false} />);
    const joinButton = getByText('Join missions');
    fireEvent.click(joinButton);
    expect(mockDispatch).toHaveBeenCalledWith(joinMission('123'));
  });

  test('dispatches leaveMission action when clicking on "Leave missions" button', () => {
    const { getByText } = render(<MissionButton id="123" reserved={true} />);
    const leaveButton = getByText('Leave missions');
    fireEvent.click(leaveButton);
    expect(mockDispatch).toHaveBeenCalledWith(leaveMission('123'));
  });
});
