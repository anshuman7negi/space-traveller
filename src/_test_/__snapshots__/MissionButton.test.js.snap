// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MissionButton Component MissionButton component renders correctly 1`] = `
<DocumentFragment>
  <button
    class="text-gray-500 bg-transparent cursor-pointer mt-3 rounded px-[1.2rem] py-2 font-medium outline outline-gray-400"
    type="button"
  >
    Join missions
  </button>
</DocumentFragment>
`;
